File: example/joint_teleop_under_lan.cpp
/**
 * @example joint_teleop_under_lan.cpp
 * @brief joint-space robot-robot teleoperation under LAN (Local Area Network) connection.
 * @copyright Copyright (C) 2016-2025 Flexiv Ltd. All Rights Reserved.
 * @author Flexiv
 */

#include <flexiv/tdk/joint_teleop_lan.hpp>

#include <spdlog/spdlog.h>

#include <getopt.h>
#include <iostream>
#include <thread>

namespace {
const std::vector<double> kJointStiffnessRatio = {0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02};
constexpr double kLastJointShapedInertia = 0.05;
}

void PrintHelp()
{
    // clang-format off
    std::cout << "Usage: sudo ./joint_teleop_under_lan [-1 serial_num] [-2 serial_num]" << std::endl;
    std::cout << "  -1  --first-sn    Serial number of the first robot." << std::endl;
    std::cout << "  -2  --second-sn   Serial number of the second robot." << std::endl;
    // clang-format on
}

const struct option kLongOptions[] = {
    // clang-format off
    {"first-sn",    required_argument, 0, '1'},
    {"second-sn",   required_argument, 0, '2'},
    {0,                             0, 0,  0}
    // clang-format on
};

int main(int argc, char* argv[])
{
    // Parse program arguments
    std::string first_sn, second_sn;
    int opt = 0;
    while ((opt = getopt_long(argc, argv, "1:2:", kLongOptions, nullptr)) != -1) {
        switch (opt) {
            case '1':
                first_sn = std::string(optarg);
                break;
            case '2':
                second_sn = std::string(optarg);
                break;
            default:
                PrintHelp();
                return 1;
        }
    }
    if (first_sn.empty() || second_sn.empty()) {
        PrintHelp();
        return 1;
    }

    try {
        // Create teleop control interface
        flexiv::tdk::JointTeleopLAN joint_teleop({{first_sn, second_sn}});

        // Only control 1 pair of robots
        unsigned int robot_pair_idx = 0;

        // Run initialization sequence
        joint_teleop.Init();

        // Set 20 degrees soft limit
        joint_teleop.SetSoftLimit(robot_pair_idx, 20.0);

        // Sync pose, first robot stays still, second robot moves to its pose
        joint_teleop.SyncPose(robot_pair_idx, {});

        // Enable inertia shaping for the last joint
        std::vector<std::pair<bool, double>> shaped_joint_inertia;
        shaped_joint_inertia.resize(joint_teleop.DoF(robot_pair_idx), {false, 1.0});
        shaped_joint_inertia.back() = {true, kLastJointShapedInertia};
        joint_teleop.SetInertiaShaping(robot_pair_idx, shaped_joint_inertia);

        // Start control loop
        joint_teleop.Start();

        // Set impedance properties
        joint_teleop.SetJointImpedance(robot_pair_idx, kJointStiffnessRatio);

        // Block until faulted
        bool last_pedal_input = false;
        while (!joint_teleop.any_fault()) {
            // Activate by pedal
            bool pedal_input = joint_teleop.digital_inputs(robot_pair_idx).first[0];
            if (pedal_input != last_pedal_input) {
                joint_teleop.Activate(robot_pair_idx, pedal_input);
                last_pedal_input = pedal_input;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }

        // Fault occurred, stop the teleoperation
        joint_teleop.Stop();

    } catch (const std::exception& e) {
        spdlog::error(e.what());
        return 1;
    }

    return 0;
}

File: example_py/joint_teleop_under_lan.py

#!/usr/bin/env python

"""
joint_teleop_under_lan.py

Example usage of joint-space robot-robot teleoperation under LAN (Local Area Network) connection.
Monitors pedal input from the first robot to activate/deactivate teleoperation.

"""

__copyright__ = "Copyright (C) 2016-2025 Flexiv Ltd. All Rights Reserved."
__author__ = "Flexiv"

import spdlog
import argparse
import time

# pip install flexivtdk
import flexivtdk

logger=spdlog.ConsoleLogger("Example")

def print_help():
    """Print usage information"""
    print("Usage: python joint_teleop_under_lan.py --first-sn <robot_sn> --second-sn <robot_sn>")
    print("  --first-sn    Serial number of the first robot.")
    print("  --second-sn   Serial number of the second robot.")


def main():
    parser = argparse.ArgumentParser(description="Joint Teleoperation over LAN")
    parser.add_argument("--first-sn", required=True, help="Serial number of the first robot")
    parser.add_argument("--second-sn", required=True, help="Serial number of the second robot")
    
    args = parser.parse_args()
    
    if not args.first_sn or not args.second_sn:
        print_help()
        return 1

    robot_pairs= [(args.first_sn, args.second_sn)]
    try:
        # Create teleop control interface
        joint_teleop = flexivtdk.JointTeleopLAN(robot_pairs)

        # Only control 1 pair of robots
        robot_pair_idx = 0

        # Joint stiffness ratios for impedance control
        joint_stiffness_ratio = [0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02]
        
        # Last joint shaped inertia value
        last_joint_shaped_inertia = 0.05

        # Run initialization sequence
        logger.info("Initializing teleoperation...")
        joint_teleop.Init()

        # Set 20 degrees soft limit
        logger.info("Setting soft limit...")
        joint_teleop.SetSoftLimit(robot_pair_idx, 20.0)

        # Sync pose, first robot stays still, second robot moves to its pose
        logger.info("Synchronizing poses...")
        joint_teleop.SyncPose(robot_pair_idx, [])

        # Enable inertia shaping for the last joint
        logger.info("Setting up inertia shaping...")
        dof = joint_teleop.DoF(robot_pair_idx)
        shaped_joint_inertia = []
        for i in range(dof):
            if i == dof - 1:  # Last joint
                shaped_joint_inertia.append((True, last_joint_shaped_inertia))
            else:
                shaped_joint_inertia.append((False, 1.0))
        
        joint_teleop.SetInertiaShaping(robot_pair_idx, shaped_joint_inertia)

        # Start control loop
        joint_teleop.Start()

        # Set impedance properties
        joint_teleop.SetJointImpedance(robot_pair_idx, joint_stiffness_ratio)

        # Store last pedal input state
        last_pedal_input = False
        
        logger.info("Teleoperation started. Press pedal to activate/deactivate.")

        # Monitor loop
        while not joint_teleop.any_fault():
            # Activate by pedal
            digital_inputs = joint_teleop.digital_inputs(robot_pair_idx)
            # First digital input of the first robot
            pedal_input = digital_inputs[0][0]  
            
            if pedal_input != last_pedal_input:
                logger.info(f"Pedal state changed: {pedal_input}")
                joint_teleop.Activate(robot_pair_idx, pedal_input)
                last_pedal_input = pedal_input
            
            # Sleep to prevent excessive CPU usage
            time.sleep(0.1)

        logger.info("Fault detected, stopping teleoperation...")
        # Fault occurred, stop the teleoperation
        joint_teleop.Stop()

    except Exception as e:
        logger.info(f"Error: {e}")
        return 1

    logger.info("Teleoperation exiting.")
    return 0


if __name__ == "__main__":
    main()